### PR TITLE
ARROW-8708: [CI] Utilize github actions cache for docker-compose volumes

### PR DESCRIPTION
**Pull request by kszucs**
Github actions has increased the repository level cache size to 5GB which can be enough for the C++ based builds. With a higher compression level the ubuntu cache consumed less than 300MB of disk. 

It also worth mentioning, that the github actions caches are evicted once a day even if the 5GB limit has been reached. So basically we don't need to worry about the cache size at all.

I set up each job to use a platform specific cache key ~prefixed with docker and postfixed with the actual git reference. This means that `docker-ubuntu-18.04-refs/pull/<pr>/merge` key is used in case of a pull request and `docker-ubuntu-18.04-master` on master. Also set a restore key to the master key as a fallback to have a cache hit on new pull requests.~

Caching has builtin support for [branch hierarchy](https://help.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows#example-of-search-priority).

### CODE REVIEW

**Comment by kszucs** [comment]
With warmed up cache the build time has been reduced to 6m from 17m which is promising.

**Comment by kszucs** [comment]
Checking that the cache properly works on my fork's master branch.

**Comment by pitrou** [comment]
Did they increase the available cache size? Last I looked it was a fixed size for the entire repo.

**Comment by kszucs** [comment]
I'm planning to merge it once turns green, because it contains a critical fix for the python workflow which was incorrectly tabulated in the previous patch.

**Comment by kszucs** [comment]
Need to trigger another CI execution to see the benefits of the caching. The previous runs showed a decent speedup, but we can certainly increase the cache hit rate later with a bit better cache key selection.

File: .github/workflows/python.yml [hunk 1]
```diff
@@ -84,6 +93,14 @@ jobs:
...
       - name: Free Up Disk Space
         run: ci/scripts/util_cleanup.sh
+      - name: Cache Docker Volumes
+        uses: actions/cache@v1
+        with:
+          path: .docker
+          key: ${{ matrix.cache }}-${{ hashFiles('cpp/**') }}
+          restore-keys: ${{ matrix.cache }}-
+      - name: Debug
+        run: mkdir -p .docker && du -d1 -h .docker
```
Comment (kou) [inline .github/workflows/python.yml:103]: Should we remove this?

File: .github/workflows/python_cron.yml [hunk 2]
```diff
@@ -110,6 +119,12 @@ jobs:
         run: ci/scripts/util_checkout.sh
       - name: Free Up Disk Space
         run: ci/scripts/util_cleanup.sh
+      - name: Cache Docker Volumes
+        uses: actions/cache@v1
+        with:
+          path: .docker
+          key: ${{ matrix.cache }}-${{ hashFiles('cpp/**') }}
```
Comment (kou) [inline .github/workflows/python_cron.yml:126]: `hashFiles('cpp/**', 'python/**')`?

**Review by kou — APPROVED** [review]
+1
I'm not sure the current `hashFiles(...)` configurations are correct...

**Comment by pitrou** [comment]
Why are you hashing the files? This means if one C++ source file changes, the ccache won't be reused?

**Comment by kszucs** [comment]
Gthub actions cache works a bit counter intuitively, once you upload something to a cache key, you cannot modify it until it gets evicted by github. Thus the keys are immutable. 

I set a restore pattern to [match previously created cache entries](https://help.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows#matching-a-cache-key) if the specific hash is not found.

**Comment by pitrou** [comment]
Yuck, I see.

File: .github/workflows/python.yml (hunk 1 above)
Comment (kszucs) [reply .github/workflows/python.yml]: Yes, removed.

File: .github/workflows/python_cron.yml (hunk 2 above)
Comment (kszucs) [reply .github/workflows/python_cron.yml:126]: The build time is majored by the cpp implementation, so I thought it's better to have cache hits if the underlying C++ implementation has not changed.

**Comment by kszucs** [comment]
Many builds have finished in around 5 minutes which is a fine improvement for now.

File: docker-compose.yml [hunk 3]
```diff
@@ -55,29 +55,12 @@
 
 version: '3.5'
 
-volumes:
```
Comment (kszucs) [inline docker-compose.yml:58]: Created a follow-up to remove the `env` boilerplates  from the github workflow files:https://issues.apache.org/jira/browse/ARROW-8713

**Pull request closed without merging**